src/azul/project/hca/transformers.py

```diff
@@ -1,17 +1,143 @@
+    processes: MutableMapping[str, Any]  # Merges process with protocol
+    files: MutableMapping[api.UUID4, Any]  # Merges manifest + file metadata
+
+    @staticmethod
+    def _merge_process_protocol(pc: api.Process, pl: api.Protocol) -> MutableMapping[str, Any]:
+        d = dict(documet_id=[pc.document_id, pl.document_id],
+                 process_id=pc.process_id,
+                 process_name=pc.process_name,
+                 protocol_id=pl.protocol_id,
+                 protocol_name=pl.protocol_name,
+                 _type="process")
+        if isinstance(pc, api.LibraryPreparationProcess):
+            d["library_construction_approach"] = pc.library_construction_approach
+        elif isinstance(pc, api.SequencingProcess):
+            d["instrument_manufacturer_model"] = pc.instrument_manufacturer_model
+        return d
+
+    @staticmethod
+    def _file_dict(f: api.File) -> MutableMapping[str, Any]:
+        d = {
+            "content-type": f.manifest_entry.content_type,
+            "crc32c": f.manifest_entry.crc32c,
+            "indexed": f.manifest_entry.indexed,
+            "name": f.manifest_entry.name,
+            "s3_etag": f.manifest_entry.s3_etag,
+            "sha1": f.manifest_entry.sha1,
+            "sha256": f.manifest_entry.sha256,
+            "size": f.manifest_entry.size,
+            "uuid": f.manifest_entry.uuid,
+            "version": f.manifest_entry.version,
+            "document_id": f.document_id,
+            "file_format": f.file_format,
+            "_type": "file"
+        }
+        if isinstance(f, api.SequenceFile):
+            d = {
+                **d,
+                "read_index": f.read_index,
+                "lane_index": f.lane_index
+            }
+        return d
+
+    def __init__(self) -> None:
+        self.specimens = {}
+        self.processes = {}
+        self.files = {}
+
+    def visit(self, entity: api.Entity) -> None:
+        if isinstance(entity, api.SpecimenFromOrganism):
+            self.specimens[entity.document_id] = entity
+        elif isinstance(entity, api.Process):
+            for pl in entity.protocols.values():
+                pl_pr_id = f"{str(pl.document_id)}-{str(entity.document_id)}"
+                self.processes[pl_pr_id] = self._merge_process_protocol(entity, pl)
+        elif isinstance(entity, api.File):
+            self.files[entity.document_id] = self._file_dict(entity)
+
+
+class BiomaterialVisitor(api.EntityVisitor):
+    specimen: MutableMapping[str, Any]
+    biomaterial_lineage: List[MutableMapping[str, Any]]
+
+    def __init__(self) -> None:
+        self.biomaterial_lineage = []
+
+    def visit(self, entity: api.Entity) -> None:
+
+        if isinstance(entity, api.Biomaterial):
+            # noinspection PyProtectedMember
+            b = dict(document_id=entity.document_id,
+                     has_input_biomaterial=entity.has_input_biomaterial,
+                     _source=entity._source)
```

Comment:
The point about protected fields is that they should not be used outside of a class or its sub-classes.